Show the same plans in the comparison grid when changing the plan term

jboland88 opened the pull request:
## Proposed Changes

* This PR allows the plan comparison grid to show the same set of plans after the plan interval has been changed. This had been an issue previously when showing fewer than the full set of plans


## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* The ideal behavior is to preserve any custom plan ordering that has been set in the comparison grid when the plan term is changed.

## Testing Instructions

* Test this PR locally or using the Calypso live link
* Navigate to the "My Home" view for one of your sites in Calypso - this is easiest to demonstrate with a free plan site
* Now, click on "Upgrades" in the left nav to view the plans page
* On the plans page, scroll down and click "Compare Plans" to show the plan comparison grid
* Reduce the size of your screen so that only 3 plans show in the comparison grid
![Screenshot 2025-06-10 at 2 28 38 PM](https://github.com/user-attachments/assets/75d34fc7-fb34-485a-ba02-838420819220)
* Use the selectors to change which plans are showing in the comparison grid
![Screenshot 2025-06-10 at 2 28 46 PM](https://github.com/user-attachments/assets/10ca0d5e-589a-4ac8-b9b1-118945027186)
* Choose plans to show in the comparison that are not the default order
![Screenshot 2025-06-10 at 2 28 54 PM](https://github.com/user-attachments/assets/3378bb8a-78de-4294-877f-e459ec6d118a)
* Now, change the billing term (for example, from one year to two years)
* Confirm that the plan prices update, and that the plans you have selected to compare do not change or switch positions
![Screenshot 2025-06-10 at 2 29 03 PM](https://github.com/user-attachments/assets/84ac59e1-4b9a-4151-85be-2f98d90e8966)
  

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

jboland88 committed:
Show the same plans in the comparison grid when changing the plan term

jboland88 committed:
Update approach to not pass state that's updated in useEffect

chriskmnds commented on packages/plans-grid-next/src/components/comparison-grid/index.tsx:
```diff
@@ -997,6 +997,12 @@ const ComparisonGrid = ( {
+	const visiblePlansRef = useRef< PlanSlug[] >( [] );
+
+	useEffect( () => {
+		visiblePlansRef.current = visiblePlans;
```
What is the need for this reference?

jboland88 replied:
Just me being rusty with state hooks and adding unnecessary code 😉

chriskmnds commented on packages/plans-grid-next/src/components/comparison-grid/index.tsx:
```diff
@@ -997,6 +997,12 @@ const ComparisonGrid = ( {
 	const [ visiblePlans, setVisiblePlans ] = useState< PlanSlug[] >( [] );
 	const currentPlanTerm = Plans.useCurrentPlanTerm( { siteId } );
 	const selectedPlanTerm = usePlanBillingPeriod( { intervalType } );
+	const prevGridPlans = useRef< GridPlan[] >( gridPlans );
```
I think there's a bit of unnecessary complexity in introducing more variables and refs to the mix. In general, it's easier to maintain data flows that are contained and not scattered across multiple variables.

I was pretty sure we fixed this issue previously, though. Somehow, it got reintroduced. I also no longer see the approach used previously, in this PR: https://github.com/Automattic/wp-calypso/pull/87017

Goes to say how easy it is to break this part of the logic over and over.

Can you please explore reintroducing the changes from https://github.com/Automattic/wp-calypso/pull/87017, which uses a callback in the state handler (`setVisiblePlans( ( prev ) => ... )`). I'd be curious about how/why we broke that along the way. We may consider abstracting it a little better to guard the logic instead.

jboland88 replied:
Ahhh, that's a helpful reference. 

This felt like it was getting a bit ugly - I appreciate the review!
I'll look at reintroducing the previous approach 👍

chriskmnds commented on packages/plans-grid-next/src/components/comparison-grid/index.tsx:
```diff
@@ -1014,9 +1020,26 @@ const ComparisonGrid = ( {
+
+		// If the number of plans is fewer than the full set and we are updating to a new plan set
+		// Show the plans at the same indexes in the new set as were shown in the old set
+		// This should get plans in the same order as the previous set of grid plans
```
Aside, inline comments aren't always helpful for maintenance. They're often a sign that the code could be structured a little better / perhaps a cleaner separation of concerns or better encapsulation is needed.

jboland88 replied:
Good note - thanks!

chriskmnds reviewed (changes requested):
Thanks for working on this! Left a few comments in the code that I feel need addressing. Aside from the bit of regression below, my main concern would be about simplifying the code a little. As this is something that was [fixed previously](https://github.com/Automattic/wp-calypso/pull/87017) (but got reintroduced in additional refactors), it would also be great to explore better abstractions to guard it more.

It appears a little broken to me. Opening the grid in a narrower view, for instance (not resizing towards there):

| Before | After |
|--------|--------|
| <img width="724" alt="Screenshot 2025-06-11 at 8 54 28 AM" src="https://github.com/user-attachments/assets/66fa6eb0-de52-4c7e-af37-fb6cfcb961d5" /> | <img width="699" alt="Screenshot 2025-06-11 at 8 54 17 AM" src="https://github.com/user-attachments/assets/16e9240f-0f63-4a52-94f3-d4fc36806a1c" /> |

jboland88 committed:
update method to use previous approach with setState callback

jboland88 committed:
Merge branch 'trunk' into update/maintain-plan-grid-order-when-switching-terms

chriskmnds reviewed (commented):
Thanks for the follow-up. :-)

Something still doesn't feel right to me. The logic isn't exactly executed as outlined. Unless I'm missing something, it's not exactly deterministic right now. I don't always see the current plan in the list, more so not as the first one from the left ("not always" as sometimes it does show). Perhaps some race condition or loopy process in the background? 🤔  We need to verify nothing like that happens.

https://github.com/user-attachments/assets/bbf0281d-a765-4e47-a6df-a0311bcfcaa6

And in the above state, when I can select the current plan anywhere in the list, it also removes the selector to set a different plan in that column - also not good:

<img width="500" alt="Screenshot 2025-06-15 at 5 36 16 PM" src="https://github.com/user-attachments/assets/9635bed3-5c4f-4e13-9333-1b424c8648a1" />

Not sure if this happens in production as well.

> It seems the one difference between then and now is an additional check to make sure the user's current plan is always shown - perhaps things broke when that was introduced 🤷‍♂️

I think the main question to be asking, from a quality standpoint, is how we ensure this (breakage) doesn't happen again. Perhaps some unit tests to cover those expectations in the comments might be a good idea? Additionally, perhaps moving the code into a separate unit might help (thinking of `useVisiblePlans` hook, but not sure). Run it through Cursor (but ask it to be critical about whatever it proposes)!

jboland88 committed:
Merge branch 'trunk' into update/maintain-plan-grid-order-when-switching-terms

jboland88 committed:
Move visible grid item logic to separate component, update sorting to favor upgrades and always show current plan at front of comparison grid

jboland88 committed:
Variable placement

jboland88 committed:
Refactor to use just the GridPlan objects in the hook w/o transform step

jboland88 committed:
add some unit tests

jboland88 committed:
Merge branch 'trunk' into update/maintain-plan-grid-order-when-switching-terms

chriskmnds commented on packages/plans-grid-next/src/components/comparison-grid/use-visible-grid-plans.tsx:
```diff
@@ -0,0 +1,157 @@
+		}
+
+		return result;
+	};
```
This looks very primitive and imperative, with mbiguous variables, multiple while loops, variable increments, etc. I'd seek a more functional equivalent. Certainly no `i`s and `n`s in the code pls. `while` loops I'd especially avoid as much as possible, particularly with complex conditioning. I don't quite see this being maintainable 🤷‍♂️

jboland88 replied:
Good point - this was some generated code that was handling which plans to return. I am going to remove it so we're not introducing extra changes beyond the scope of this PR

chriskmnds commented on packages/plans-grid-next/src/components/comparison-grid/use-visible-grid-plans.tsx:
```diff
@@ -0,0 +1,157 @@
+				} );
+			}
+
+			// Ensure current plan is visible and at the start
```
I'm curious if we start with the question "when would the current plan not be visible and the first item in the list?" whether it might lead to a different form of processing here. It's complex block of code.

jboland88 replied:
Good question. switching this back to the previous logic where it just makes sure the plan is always visible, but not necessarily first simplifies things quite a bit. I am going to go back in that direction

chriskmnds commented on packages/plans-grid-next/src/components/comparison-grid/use-visible-grid-plans.tsx:
```diff
@@ -0,0 +1,157 @@
+			// Check if previous state is stale
+			const isPrevStale = prev.some( ( plan ) => ! gridPlansIndex[ plan.planSlug ] );
+
+			let next: GridPlan[] = prev;
```
In general, I think the more we can avoid mutations, the better. The alternatives are sometimes better functional code. As long as we don't end up in a series of early returns (also not great). 

Also, names like "next" and "prev" are a little too generic - even `prevState` and `nextState` would be better.

chriskmnds replied:
> As long as we don't end up in a series of early returns (also not great).

Nothing against early returns. As long as it's not overdone.

jboland88 replied:
Interesting, I'm curious how we'd avoid a mutable variable here with two separate steps that are dependent - one to determine the initial set of plans and another to slice the current plan into the array if missing 🤔

chriskmnds commented on packages/plans-grid-next/src/components/comparison-grid/use-visible-grid-plans.tsx:
```diff
@@ -0,0 +1,157 @@
+			}
+		},
+		[ gridPlans ]
+	);
```
This is a concern (event handler) of the main component, so I think best for it to live there. I could be wrong though. We can return a `setVisibleGridPlans` setter from here as the alternative.

jboland88 replied:
good call 👍

chriskmnds commented on packages/plans-grid-next/src/components/comparison-grid/use-visible-grid-plans.tsx:
Let's move this to the `/hooks` folder

chriskmnds reviewed (commented):
Heya. Thanks for following up. It looks closer, but I think it needs a bit more work. The code looks a little complex, which makes me wonder if we are aiming for a wider target than necessary at places. In terms of switching the term and interacting with the toggles, it works as expected with selections maintained (except for when resizing, the order is lost, but it's also that way in production) 

A few more thoughts:

- This now introduces a different ordering than what is in production. Not sure if this is impromptu or something that is known to be broken. In the spirit of keeping PRs focused with narrow scope, I'd rather we move this to a separate PR (and issue). It sounds wider in scope than "show the same plans when changing term".
- There is also what looks like a different breakpoint on which the current plan is moved to the left (before/after below). There may be value in keeping the order ascending in tier/price when everything's displayed (so as in production), but maybe not. I wonder if it was intentional that way. In any case, I think worth of separate PR.

### Before
 <img width="1240" alt="Screenshot 2025-06-18 at 8 30 21 PM" src="https://github.com/user-attachments/assets/fc41bff0-77aa-43e2-8182-c30589256902" /> 

### After
<img width="1249" alt="Screenshot 2025-06-18 at 8 30 12 PM" src="https://github.com/user-attachments/assets/1e4aea4d-d330-43cd-9ecf-b2209dd86041" />

jboland88 committed:
Merge branch 'trunk' into update/maintain-plan-grid-order-when-switching-terms

jboland88 committed:
Updates from feedback

jboland88 committed:
rename variable for clarity

jboland88 commented:
Thanks again for more feedback @chriskmnds - I have been improved by revisiting and making updates!

Indeed, I had become a bit too eager and introduced some self-labeled "improvements" (different sort order, forcing current plan to always be first). I've removed these as they were out of scope for sure.

The hook is now a bit simpler and the event handler is back in the correct context 👍 
Let's see if it passes the Christos test now 😉

jboland88 committed:
Merge branch 'trunk' into update/maintain-plan-grid-order-when-switching-terms

chriskmnds commented:
Sorry if I left you in cold water here, Josh. I completely neglected this.

Do you still need input from me? I will give it a post-merge review :-)
